content/blog/2019/11/2019-11-21-welcome-to-the-matrix.adoc

```diff
@@ -0,0 +1,389 @@
+                        axis {
+                            name 'BROWSER'
+                            values 'edge'
+                        }
+                    }
+                }
+                stages {
+                    stage('Build') {
+                        steps {
+                            echo "Do Build for ${PLATFORM} - ${BROWSER}"
+                        }
+                    }
+                    stage('Test') {
+                        steps {
+                            echo "Do Test for ${PLATFORM} - ${BROWSER}"
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+----
+
+.Log output (truncated)
+[source]
+----
+...
+[Pipeline] stage
+[Pipeline] { (BuildAndTest)
+[Pipeline] parallel
+[Pipeline] { (Branch: Matrix - OS = 'linux', BROWSER = 'firefox')
+[Pipeline] { (Branch: Matrix - OS = 'windows', BROWSER = 'firefox')
+[Pipeline] { (Branch: Matrix - OS = 'mac', BROWSER = 'firefox')
+[Pipeline] { (Branch: Matrix - OS = 'linux', BROWSER = 'chrome')
+[Pipeline] { (Branch: Matrix - OS = 'windows', BROWSER = 'chrome')
+[Pipeline] { (Branch: Matrix - OS = 'mac', BROWSER = 'chrome')
+[Pipeline] { (Branch: Matrix - OS = 'windows', BROWSER = 'safari')
+[Pipeline] { (Branch: Matrix - OS = 'mac', BROWSER = 'safari')
+[Pipeline] { (Branch: Matrix - OS = 'windows', BROWSER = 'edge')
+...
+Do Build for linux - firefox
+...
+----
+
+== Controlling cell behavior at runtime
+
+Inside the `matrix` directive I can also add "per-cell" directives. 
+These are the same directives that I would add to a `stage` and they let me control the behavior of each cell in the matrix.
+These directives can use the axis values from their cell as part of their inputs, allowing me to customize the behavior of each cell to match its axis values. 
+
+On my Jenkins server I have configured agents with labels that match the OS for each agent ("linux-agent", "windows-agent", and "mac-agent"). 
+To run each cell in my matrix on the appropriate operating system, I configure the label for that cell using Groovy string templating. 
+
+[source, groovy]
+----
+matrix {
+    axes { ... }
+    excludes { ... }
+    agent {
+        label "${PLATFORM}-agent"
+    }
+    stages { ... }
+    // ...
+}
+----
+
+
+Occasionally I run my pipeline manually from the Jenkins Web UI. 
+When I do that, I'd like to be able to select just one platform to run, but I can't use `excludes` for this.
+The `axis` and `exclude` directives define the static set of cells that make up the matrix. 
+That set of combinations is generated before the start of the run, before any parameters. 
+
+So, I can't add or remove cells from a matrix after the job has started but I can use a "per-cell" directives. 
+I'll add a `choice` parameter with the list of platforms and use a `when` directive inside `matrix` to control which cells in the matrix are run.
+The log below shows a portion of the output when I select `PLATFORM_FILTER = 'mac'`.
```

Comment:
If you do something like what I've described above, this line could be moved so that it's right on top of the sample log output, rather than being up here.